File: examples/node/pwa-express/template.js
/**
 * Template and response
 * See: https://github.com/FranBarInstance/neutralts-docs
 */

const NeutralIpcTemplate = require('../neutral_ipc_template/NeutralIpcTemplate');

const {
    TEMPLATE_ROUTER,
    TEMPLATE_ERROR
} = require('./constants');

class Template {
    constructor(schema) {
        this.schema = schema;
    }

    async render(req, res) {
        let template = new NeutralIpcTemplate(TEMPLATE_ROUTER, JSON.stringify(this.schema));
        let contents = await template.render();
        let statusCode = template.getStatusCode();
        let statusText = template.getStatusText();
        let statusParam = template.getStatusParam();

        // The template may generate redirects.
        if (statusCode && [301, 302, 307, 308].includes(parseInt(statusCode))) {
            return res.redirect(parseInt(statusCode), statusParam);
        }

        // The template may generate HTTP errors.
        if (statusCode && parseInt(statusCode) >= 400) {
            const error = {
                data: {
                    CONTEXT: {
                        ROUTE: 'error'
                    },
                    error: {
                        code: parseInt(statusCode),
                        text: statusText,
                        param: statusParam
                    }
                }
            };
            template.setPath(TEMPLATE_ERROR);
            template.mergeSchema(JSON.stringify(error));
            contents = await template.render();
        }

        return res.status(parseInt(statusCode) || 200).type('text/html').send(contents);
    }
}

module.exports = Template;

File: examples/rust/pwa-actix-ipc/src/template.rs
//! Template and response
//! See: https://github.com/FranBarInstance/neutralts-docs

use actix_web::{HttpResponse, http::StatusCode};
use neutralipcrs::NeutralIpcTemplate;
use serde_json::{json, Value};
use crate::constants::{TEMPLATE_ERROR, TEMPLATE_ROUTER};

pub struct Template {
    schema: Value,
}

impl Template {
    pub fn new(schema: Value) -> Self {
        Self { schema }
    }

    pub fn render(&self) -> Result<HttpResponse, Box<dyn std::error::Error>> {

        // Create template from file with schema
        let mut template = NeutralIpcTemplate::from_file_value(TEMPLATE_ROUTER, self.schema.clone())?;
        let contents = template.render().unwrap();

        // Get the status code.
        // Neutral TS may send an HTTP error code, with {:redirect; ... :} or {:exit; ... :}
        let status_code = template.get_status_code();
        let status_text = template.get_status_text();
        let status_param = template.get_status_param();

        if let Ok(code) = status_code.parse::<u16>() {

            // The template may generate redirects.
            if [301, 302, 307, 308].contains(&code) {
                return Ok(HttpResponse::build(StatusCode::from_u16(code).unwrap())
                    .append_header(("Location", status_param.to_string()))
                    .finish());
            }

            // The template may generate HTTP errors.
            if code >= 400 {
                let error = json!({
                    "data": {
                        "CONTEXT": {
                            "ROUTE": "error"
                        },
                        "error": {
                            "code": code,
                            "text": status_text,
                            "param": status_param
                        }
                    }
                });

                // Again with error page
                template.set_path(TEMPLATE_ERROR);

                // Error variables are added to the schema.
                let _ = template.merge_schema(error);

                // Render the error page
                let error_contents = template.render().unwrap();

                return Ok(HttpResponse::build(StatusCode::from_u16(code).unwrap())
                    .content_type("text/html")
                    .body(error_contents));
            }
        }

        // 200 Ok
        Ok(HttpResponse::Ok()
            .content_type("text/html")
            .body(contents))
    }
}
